src/IB/IBFESurfaceMethod.cpp

```diff
@@ -1648,127 +4170,609 @@ IBFESurfaceMethod::imposeJumpConditions(const int f_data_idx,
-                            std::vector<libMesh::Point> ref_coords(1, xi);
-                            fe->reinit(elem, &ref_coords);
-                            for (unsigned int l = 0; l < NDIM - 1; ++l)
-                            {
-                                interpolate(dx_dxi[l], 0, x_node, *dphi_dxi[l]);
-                            }
-                            if (NDIM == 2)
+                            const libMesh::Point x = r + intersections[k].first * q;
+                            const libMesh::Point& xi = intersections[k].second;
+                            SideIndex<NDIM> i_s(i_c, axis, 0);
+                            i_s(axis) = boost::math::iround((x(axis) - x_lower[axis]) / dx[axis]) + patch_lower[axis];
+                            if (extended_box.contains(i_s))
                             {
-                                dx_dxi[1] = VectorValue<double>(0.0, 0.0, 1.0);
+                                std::vector<libMesh::Point> ref_coords(1, xi);
+                                fe->reinit(elem, &ref_coords);
+                             
+                                fe_interpolator.reinit(elem);
+								fe_interpolator.collectDataForInterpolation(elem);
+								fe_interpolator.interpolate(elem);
+	
+								
+								interpolate(XX, 0, X_node, phi);
+								interpolate(xx, 0, x_node, phi);
+
+                                for (unsigned int l = 0; l < NDIM - 1; ++l)
+                                {
+                                    interpolate(dx_dxi[l], 0, x_node, *dphi_dxi[l]);
+                                    interpolate(dX_dxi[l], 0, X_node, *dphi_dxi[l]);
+                                }
+                                if (NDIM == 2)
+                                {
+									dX_dxi[1] = VectorValue<double>(0.0, 0.0, 1.0);
+                                    dx_dxi[1] = VectorValue<double>(0.0, 0.0, 1.0);
+                                }
+                                N = dX_dxi[0].cross(dX_dxi[1]);
+                                const double dA = N.norm();
+                                N = N.unit();
+                                n = dx_dxi[0].cross(dx_dxi[1]);
+                                const double da = n.norm();
+                                n = n.unit();
+                                
+                
+
+                                // Make sure we haven't already found this
+                                // intersection.
+                                //
+                                // (Because we are doing this in floating point
+                                // arithmetic, we can't even count on the
+                                // intersection being assigned to the same index!)
+                                bool found_same_intersection_point = false;
+                                for (int shift = -1; shift <= 1; ++shift)
+                                {
+                                    SideIndex<NDIM> i_s_prime = i_s;
```

Comment:
I think this is nearly the same, or the same, intersection code as above. It should be in a separate function if possible.